runtime/lua/lsp/client.lua

```diff
@@ -0,0 +1,400 @@
+
+  return self
+end
+
+client.initialize = function(self)
+  -- Initialize autocmd for filetypes == self.ft
+  --    These will create buffer local autocmds in each buffer
+  initialize_filetype_autocmds(self.ft)
+
+  local result = self:request_async('initialize', nil, function(_, data)
+    self.capabilities =  EmptyDictionary:new(data.capabilities)
+    return data.capabilities
+  end)
+
+  self:notify('initialized')
+
+  return result
+end
+
+client.close = function(self)
+  if self._closed then
+    return
+  end
+
+  self._closed = true
+  vim.api.nvim_call_function('jobclose', {self.job_id})
+end
+
+--- Make a request to the server
+-- @param method: Name of the LSP method
+-- @param params: the parameters to send
+-- @param cb (optional): If sent, will call this when it's done
+--                          otherwise it'll wait til the client is done
+client.request = function(self, method, params, cb)
+  if not method then
+    return nil
+  end
+
+  local request_id = self:request_async(method, params, cb)
+
+  -- local later = os.time() + require('lsp.conf.request').timeout
+  local later = os.time() + 10
+
+  while (os.time() < later) and (self._results[request_id]  == nil) do
+    vim.api.nvim_command('sleep 100m')
+  end
+
+  if self._results[request_id] == nil then
+    return nil
+  end
+
+  return self._results[request_id].result[1]
+end
+
+--- Sends an async request to the client.
+-- If a callback is passed,
+--  it will be registered to run when the response is received.
+--
+-- It will also issue autocommands in Vim so that other plugins can do things on LSP actions.
+--  For example, when sending the request: 'textDocument/hover',
+--  it will send:
+--      LSP/textDocument/hover/pre
+--      LSP/textDocument/hover/post
+--
+-- @param method (string|table) : The identifier for the type of message that is being requested
+-- @param params (table)        : Optional parameters to pass to override default parameters for a request
+-- @param cb     (function)     : An optional function pointer to call once the request has been completed
+--                                  If a string is passed, it will execute a VimL funciton of that name
+--                                  To disable handling the request, pass "false"
+client.request_async = function(self, method, params, cb)
+  if self.job_id == nil then
+    log.warn('Client does not have valid job_id: ', self.name)
+    return nil
+  end
+
+  local req = message.RequestMessage:new(self, method, params)
+
+  if req == nil then
+    return nil
+  end
+
+  -- After handling callback semantics, store it to call on reply.
+  if cb ~= false then
+    self._callbacks[req.id] = {
+      cb = cb,
+      method = req.method,
+    }
+  end
+
+
+  if should_send_message(self, req) then
+    lsp_doautocmd(method, 'pre')
+    log.debug("Sending Request: [["..req:data().."]]")
+    vim.api.nvim_call_function('chansend', {self.job_id, req:data()})
+    lsp_doautocmd(method, 'post')
```

Comment:
I don't know what we really need 'pre' and 'post' hooks.

Reply:
I think we should very much provide those. Imho essential for extending functionality, which is kind of the point of an "lsp library".